pkg/scheduler/scheduler.go

```diff
@@ -764,6 +762,23 @@ func (sched *Scheduler) scheduleOne(ctx context.Context) {
 	}()
 }
 
+// skipPodSchedule returns true if we could skip scheduling the pod for specified cases.
+func (sched *Scheduler) skipPodSchedule(pod *v1.Pod) bool {
+	// Case 1: pod is being deleted.
+	if pod.DeletionTimestamp != nil {
+		sched.Recorder.Eventf(pod, nil, v1.EventTypeWarning, "FailedScheduling", "Scheduling", "skip schedule deleting pod: %v/%v", pod.Namespace, pod.Name)
+		klog.V(3).Infof("Skip schedule deleting pod: %v/%v", pod.Namespace, pod.Name)
+		return true
+	}
+
+	// Case 2: pod has been assumed and pod updates could be skipped.
+	if sched.skipPodUpdate(pod) {
```

Comment:
This looks strange because ```sched.skipPodUpdate(pod)``` is checking if the pod is ```DeepEqual``` with itself (which will always be true).

As you mention in the comment, this is basically checking if the pod assumed or not, which negates the logic in eventhandlers.go for accepting pod update events: here you are skipping all assumed pods, but the update event handler skips assumed pods _that didn't change_.

I think the main issue here is that skipPodUpdate does not skip updates for unassumed pods even if they didn't change. So the question that we need to answer is this: can we skip updates for unchanged pods even if they were not assumed? are there cases that prevents us from doing so?

Reply:
> This looks strange because sched.skipPodUpdate(pod) is checking if the pod is DeepEqual with itself (which will always be true).

`sched.skipPodUpdate` is defined as following. https://github.com/kubernetes/kubernetes/blob/26d491f413e4fe992209e8d68b40d0048ebb2698/pkg/scheduler/eventhandlers.go#L286-L291

The pod is compared with the assumed pod which is cached in `schedulerCache.assumedPods`, so it is not checking pods with itself.

> As you mention in the comment, this is basically checking if the pod assumed or not, which negates the logic in eventhandlers.go for accepting pod update events: here you are skipping all assumed pods, but the update event handler skips assumed pods that didn't change.

It does not skip all assumed pods as said above, and eventhandlers.go also calls `sched.skipPodUpdate(pod)` to skip unneeded updates at [updatePodInSchedulingQueue](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/eventhandlers.go#L176).

> I think the main issue here is that skipPodUpdate does not skip updates for unassumed pods even if they didn't change. So the question that we need to answer is this: can we skip updates for unchanged pods even if they were not assumed? are there cases that prevents us from doing so?

Actually the pod's annotation is changed(which is described at https://github.com/kubernetes/kubernetes/issues/86198), howerver if the pod has been assumed(this case has been resolved in https://github.com/kubernetes/kubernetes/issues/52914) or is being scheduled(I want to address it in this PR), scheduler could skip scheduling it again.  Why we do not skip these updates if the pod is not in above cases? I guess the annotation change might be used by some scheduler plugins, so we do not.

Reply:
> > This looks strange because sched.skipPodUpdate(pod) is checking if the pod is DeepEqual with itself (which will always be true).
> 
> `sched.skipPodUpdate` is defined as following.
> 
> https://github.com/kubernetes/kubernetes/blob/26d491f413e4fe992209e8d68b40d0048ebb2698/pkg/scheduler/eventhandlers.go#L286-L291
> 
> The pod is compared with the assumed pod which is cached in `schedulerCache.assumedPods`, so it is not checking pods with itself.
> 
> > As you mention in the comment, this is basically checking if the pod assumed or not, which negates the logic in eventhandlers.go for accepting pod update events: here you are skipping all assumed pods, but the update event handler skips assumed pods that didn't change.
> 
> It does not skip all assumed pods as said above, and eventhandlers.go also calls `sched.skipPodUpdate(pod)` to skip unneeded updates at [updatePodInSchedulingQueue](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/eventhandlers.go#L176).
> 

Right, I missed that we make a deep copy of the pod instance on assume.

> > I think the main issue here is that skipPodUpdate does not skip updates for unassumed pods even if they didn't change. So the question that we need to answer is this: can we skip updates for unchanged pods even if they were not assumed? are there cases that prevents us from doing so?
> 
> Actually the pod's annotation is changed(which is described at #86198), howerver if the pod has been assumed(this case has been resolved in #52914) or is being scheduled(I want to address it in this PR), scheduler could skip scheduling it again. Why we do not skip these updates if the pod is not in above cases? I guess the annotation change might be used by some scheduler plugins, so we do not.

What I was trying to suggest is to find a way to prevent the pod from being added to the queue in the first place for the issue described in  #86198 (e.g., not add a pod that is currently being scheduled to the queue), but it might not be straight forward to achieve.

In any case, can you add a comment explaining when we hit this case: "An assumed pod can be added again to the scheduling queue if it got an update event during its previous scheduling cycle but before getting assumed."

Reply:
> What I was trying to suggest is to find a way to prevent the pod from being added to the queue in the first place for the issue described in #86198 (e.g., not add a pod that is currently being scheduled to the queue), but it might not be straight forward to achieve.

It will be great to avoid it in the first place :) I tried to get/record the pod that is being scheduled, howerver as you said, it is not straight forward, e.g. we could add an field that records the pod being scheduled, howerver it does not look good.

Reply:
> In any case, can you add a comment explaining when we hit this case: "An assumed pod can be added again to the scheduling queue if it got an update event during its previous scheduling cycle but before getting assumed."

Addressed. PTAL

Reply:
On second thought, it is not easy to deal with it in the first place. When it is being scheduled, there might not feasible nodes for it, and it will not get assumed, then we do not need to skip the upate.
So I think here is the right place to address it.